From: IrfanMohammad <irfan@graphiant.com>
Date: Thu, 27 Jul 2023 12:13:45 +0000
Subject: Cache sr_mod_oper_sub_t shm during pull callback

 To access the xpath of the oper sub, we held on to ext remap read lock
 which is not ideal, as it leads to a deadlock as described in ENG-3841.

 The relevant steps here for operational pull data are

 Step 1. oper sub lock read
 Step 2. ext remap lock read, read the evpipe, xpath of the oper sub
 Step 3. invoke callback
   Step 3a. sr_process_events (sr_module_oper_data_dup_enabled) ext WRITE lock
   ^^^ DEADLOCKED ^^^
   Step 3b. sr_process_events ext WRITE unlock
 Step 4. ext remap unlock read
 Step 5. oper sub unlock read

 Normally Step 3a requires only a read lock. However, if ext shm size
 has changed, due to another subscription add/del, we will need to remap it.
 This will require a write lock, and since read lock is already held in
 Step 2 this deadlocks.

 This can be prevented by caching `sr_mod_oper_sub_t` and unlocking
 ext_shm in Step 2. With this change the behavior is

 Step 1. oper sub lock read
 Step 2. ext remap lock read, cache mod_oper_sub info, unlock ext shm
 Step 3. invoke callback
   Step 3a. sr_process_events (sr_module_oper_data_dup_enabled) ext WRITE lock
   Step 3b. sr_process_events ext WRITE unlock
 Step 4. oper sub unlock read

Also fix wrong assumption in new_wrlock about cleared events

cherry-picked from 0068-Cache-sr_mod_oper_sub_t-shm-during-pull-callback.patch
---
 src/modinfo.c   | 81 +++++++++++++++++++++++++++++++++++++++------------------
 src/shm_ext.c   | 41 ++++++++++-------------------
 src/shm_ext.h   |  7 +++--
 src/shm_sub.c   | 23 ++++++++++------
 src/shm_sub.h   |  2 +-
 src/shm_types.h |  9 +++++++
 6 files changed, 98 insertions(+), 65 deletions(-)

diff --git a/src/modinfo.c b/src/modinfo.c
index 298ddd0..4859e2c 100644
--- a/src/modinfo.c
+++ b/src/modinfo.c
@@ -895,7 +895,7 @@ cleanup:
  */
 static sr_error_info_t *
 sr_xpath_oper_data_get(struct sr_mod_info_mod_s *mod, const char *xpath, const char **request_xpaths,
-        uint32_t req_xpath_count, const char *orig_name, const void *orig_data, sr_mod_oper_get_sub_t *shm_subs,
+        uint32_t req_xpath_count, const char *orig_name, const void *orig_data, sr_mod_oper_get_sub_cached_t *cached_subs,
         uint32_t idx1, const struct lyd_node *parent, uint32_t timeout_ms, sr_conn_ctx_t *conn,
         struct lyd_node **oper_data)
 {
@@ -940,7 +940,7 @@ sr_xpath_oper_data_get(struct sr_mod_info_mod_s *mod, const char *xpath, const c
     request_xpath = (req_xpath_count == 1) ? request_xpaths[0] : NULL;
 
     /* get data from client */
-    if ((err_info = sr_shmsub_oper_get_notify(mod, xpath, request_xpath, parent_dup, orig_name, orig_data, shm_subs,
+    if ((err_info = sr_shmsub_oper_get_notify(mod, xpath, request_xpath, parent_dup, orig_name, orig_data, cached_subs,
             idx1, timeout_ms, conn, oper_data, &cb_err_info))) {
         sr_errinfo_merge(&err_info, cb_err_info);
         goto cleanup;
@@ -1028,6 +1028,17 @@ cleanup:
     return err_info;
 }
 
+static void
+sr_mod_free_cached_sub(sr_mod_oper_get_sub_cached_t *subs, size_t count)
+{
+    size_t i;
+    for (i = 0; i < count; i++) {
+        free(subs[i].xpath);
+        free(subs[i].xpath_subs);
+    }
+    free(subs);
+}
+
 /**
  * @brief Update (replace or append) operational data for a specific module.
  *
@@ -1054,6 +1065,8 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
     int required, merged;
     struct ly_set *set = NULL;
     struct lyd_node *edit = NULL, *oper_data;
+    sr_mod_oper_get_sub_cached_t *cached_subs = NULL;
+    size_t oper_get_sub_count = 0, count = 0;
 
     if (!(get_oper_opts & SR_OPER_NO_STORED)) {
         /* get stored operational edit */
@@ -1090,31 +1103,52 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
 
     /* EXT READ LOCK */
     if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_READ, 0, __func__))) {
-        goto cleanup_opergetsub_unlock;
+        goto cleanup;
     }
+    oper_get_sub_count = mod->shm_mod->oper_get_sub_count;
+    cached_subs = calloc(oper_get_sub_count, sizeof *cached_subs);
 
     /* XPaths are ordered based on depth */
     shm_subs = (sr_mod_oper_get_sub_t *)(conn->ext_shm.addr + mod->shm_mod->oper_get_subs);
-    for (i = 0; i < mod->shm_mod->oper_get_sub_count; ++i) {
-        sub_xpath = conn->ext_shm.addr + shm_subs[i].xpath;
-        xpath_subs = (sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + shm_subs[i].xpath_subs);
 
+    for (i = 0; i < oper_get_sub_count; ++i) {
         /* useless to retrieve configuration data, state data */
         if (((shm_subs[i].sub_type == SR_OPER_GET_SUB_CONFIG) && (get_oper_opts & SR_OPER_NO_CONFIG)) ||
                 ((shm_subs[i].sub_type == SR_OPER_GET_SUB_STATE) && (get_oper_opts & SR_OPER_NO_STATE))) {
             continue;
         }
+        count = shm_subs[i].xpath_sub_count;
+        assert(conn->ext_shm.addr + shm_subs[i].xpath);
+        cached_subs[i].xpath = strdup(conn->ext_shm.addr + shm_subs[i].xpath);
+        cached_subs[i].xpath_subs = malloc(count * sizeof *cached_subs[i].xpath_subs);
+        memcpy(cached_subs[i].xpath_subs, conn->ext_shm.addr + shm_subs[i].xpath_subs, count * sizeof *cached_subs[i].xpath_subs);
+        cached_subs[i].xpath_sub_count = count;
+    }
+
+    /* prevent use after ext remap unlock */
+    shm_subs = NULL;
+
+    /* EXT READ UNLOCK */
+    sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 0, __func__);
+
+    for (i = 0; i < oper_get_sub_count; ++i) {
+        sub_xpath = cached_subs[i].xpath;
+        xpath_subs = cached_subs[i].xpath_subs;
+        if (!xpath_subs) {
+            /* Previously filtered */
+            continue;
+        }
 
         if (mod->xpath_count) {
             /* check whether these data are even required */
             for (j = 0; j < mod->xpath_count; ++j) {
                 if ((err_info = sr_xpath_oper_data_required(mod->xpaths[j], sub_xpath, &required))) {
-                    goto cleanup_opergetsub_ext_unlock;
+                    goto cleanup;
                 }
                 if (required) {
                     /* remember all xpaths causing these data to be required */
                     request_xpaths = sr_realloc(request_xpaths, (req_xpath_count + 1) * sizeof *request_xpaths);
-                    SR_CHECK_MEM_GOTO(!request_xpaths, err_info, cleanup_opergetsub_ext_unlock);
+                    SR_CHECK_MEM_GOTO(!request_xpaths, err_info, cleanup);
                     request_xpaths[req_xpath_count] = mod->xpaths[j];
                     ++req_xpath_count;
                 }
@@ -1128,13 +1162,13 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
 
         /* remove any present data */
         if (!(xpath_subs[0].opts & SR_SUBSCR_OPER_MERGE) && (err_info = sr_lyd_xpath_complement(data, sub_xpath))) {
-            goto cleanup_opergetsub_ext_unlock;
+            goto cleanup;
         }
 
         if (!(get_oper_opts & SR_OPER_NO_POLL_CACHED)) {
             /* try to get data from the cache */
             if ((err_info = sr_module_oper_data_update_cached(mod, sub_xpath, conn, data, &merged))) {
-                goto cleanup_opergetsub_ext_unlock;
+                goto cleanup;
             }
             if (merged) {
                 /* we have the data */
@@ -1144,7 +1178,7 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
 
         /* trim the last node to get the parent */
         if ((err_info = sr_xpath_trim_last_node(sub_xpath, &parent_xpath))) {
-            goto cleanup_opergetsub_ext_unlock;
+            goto cleanup;
         }
 
         if (parent_xpath) {
@@ -1154,7 +1188,7 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
             }
 
             if ((err_info = sr_lyd_find_xpath(*data, parent_xpath, &set))) {
-                goto cleanup_opergetsub_ext_unlock;
+                goto cleanup;
             }
 
             if (!set->count) {
@@ -1166,21 +1200,21 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
             if (xpath_subs[0].opts & SR_SUBSCR_OPER_ONCE) {
                 /* get oper data from the client */
                 if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                        orig_data, shm_subs, i, NULL, timeout_ms, conn, &oper_data))) {
-                    goto cleanup_opergetsub_ext_unlock;
+                        orig_data, cached_subs, i, NULL, timeout_ms, conn, &oper_data))) {
+                    goto cleanup;
                 }
 
                 /* merge into one data tree */
                 if ((err_info = sr_lyd_merge(data, oper_data, 1, LYD_MERGE_DESTRUCT))) {
                     lyd_free_all(oper_data);
-                    goto cleanup_opergetsub_ext_unlock;
+                    goto cleanup;
                 }
             } else {
                 for (j = 0; j < set->count; ++j) {
                     /* get oper data from the client */
                     if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                            orig_data, shm_subs, i, set->dnodes[j], timeout_ms, conn, &oper_data))) {
-                        goto cleanup_opergetsub_ext_unlock;
+                            orig_data, cached_subs, i, set->dnodes[j], timeout_ms, conn, &oper_data))) {
+                        goto cleanup;
                     }
 
                     /* merge into one data tree */
@@ -1200,13 +1234,13 @@ next_iter:
         } else {
             /* top-level data */
             if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                    orig_data, shm_subs, i, NULL, timeout_ms, conn, &oper_data))) {
-                goto cleanup_opergetsub_ext_unlock;
+                    orig_data, cached_subs, i, NULL, timeout_ms, conn, &oper_data))) {
+                goto cleanup;
             }
 
             if ((err_info = sr_lyd_merge(data, oper_data, 1, LYD_MERGE_DESTRUCT))) {
                 lyd_free_all(oper_data);
-                goto cleanup_opergetsub_ext_unlock;
+                goto cleanup;
             }
         }
 
@@ -1215,14 +1249,11 @@ next_iter:
         req_xpath_count = 0;
     }
 
-cleanup_opergetsub_ext_unlock:
-    /* EXT READ UNLOCK */
-    sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 0, __func__);
-
-cleanup_opergetsub_unlock:
+cleanup:
     /* OPER GET SUB READ UNLOCK */
     sr_rwunlock(&mod->shm_mod->oper_get_lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid, __func__);
 
+    sr_mod_free_cached_sub(cached_subs, oper_get_sub_count);
     free(request_xpaths);
     free(parent_xpath);
     ly_set_free(set, NULL);
diff --git a/src/shm_ext.c b/src/shm_ext.c
index 6ba38bd..25be6e6 100644
--- a/src/shm_ext.c
+++ b/src/shm_ext.c
@@ -892,7 +892,7 @@ sr_shmext_oper_get_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t sub_
             xpath_sub = &((sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + shm_sub->xpath_subs))[j];
             if (!sr_conn_is_alive(xpath_sub->cid)) {
                 /* subscription is dead, recover it */
-                if ((err_info = sr_shmext_oper_get_sub_stop(conn, shm_mod, i, j, 1, SR_LOCK_WRITE, 1))) {
+                if ((err_info = sr_shmext_oper_get_sub_stop(conn, shm_mod, i, j, SR_LOCK_WRITE, 0))) {
                     goto cleanup_unlock;
                 }
             } else {
@@ -1094,28 +1094,21 @@ cleanup_unlock:
 }
 
 sr_error_info_t *
-sr_shmext_oper_get_sub_stop(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t del_idx1, uint32_t del_idx2, int del_evpipe,
-        sr_lock_mode_t has_locks, int recovery)
+sr_shmext_oper_get_sub_stop(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t del_idx1, uint32_t del_idx2,
+        sr_lock_mode_t has_locks, uint32_t sub_id)
 {
     sr_error_info_t *err_info = NULL, *tmp_err;
     sr_mod_oper_get_sub_t *shm_sub;
     sr_mod_oper_get_xpath_sub_t *xpath_sub;
     char *path;
-    uint32_t evpipe_num, sub_id;
+    uint32_t evpipe_num;
+    int del_evpipe = 1;
 
-    assert((has_locks == SR_LOCK_READ) || (has_locks == SR_LOCK_WRITE));
+    assert((has_locks == SR_LOCK_NONE) || (has_locks == SR_LOCK_WRITE));
 
     /* get sub write lock keeping the lock order */
 
-    if (has_locks == SR_LOCK_READ) {
-        /* remember also the sub ID */
-        shm_sub = (sr_mod_oper_get_sub_t *)(conn->ext_shm.addr + shm_mod->oper_get_subs);
-        xpath_sub = (sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + shm_sub[del_idx1].xpath_subs);
-        sub_id = xpath_sub[del_idx2].sub_id;
-
-        /* EXT READ UNLOCK */
-        sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 0, __func__);
-
+    if (has_locks == SR_LOCK_NONE) {
         /* OPER GET SUB READ UNLOCK */
         sr_rwunlock(&shm_mod->oper_get_lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid, __func__);
 
@@ -1134,20 +1127,19 @@ sr_shmext_oper_get_sub_stop(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t del
     shm_sub = (sr_mod_oper_get_sub_t *)(conn->ext_shm.addr + shm_mod->oper_get_subs);
 
     /* check that the subscription has not been removed while we were not holding the locks */
-    if ((has_locks == SR_LOCK_READ) && (shm_mod->oper_get_sub_count <= del_idx1)) {
+    if ((has_locks == SR_LOCK_NONE) && (shm_mod->oper_get_sub_count <= del_idx1)) {
         xpath_sub = NULL;
     } else {
         xpath_sub = (sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + shm_sub[del_idx1].xpath_subs);
     }
-    if ((has_locks == SR_LOCK_READ) && (!xpath_sub || (shm_sub[del_idx1].xpath_sub_count <= del_idx2) ||
+
+    if ((has_locks == SR_LOCK_NONE) && (!xpath_sub || (shm_sub[del_idx1].xpath_sub_count <= del_idx2) ||
             (xpath_sub[del_idx2].sub_id != sub_id))) {
         /* already removed, do nothing */
         del_evpipe = 0;
     } else {
-        if (recovery) {
-            SR_LOG_WRN("Recovering module \"%s\" operational get subscription of CID %" PRIu32 ".",
-                    conn->mod_shm.addr + shm_mod->name, xpath_sub[del_idx2].cid);
-        }
+        SR_LOG_WRN("Recovering module \"%s\" operational get subscription of CID %" PRIu32 ".",
+                conn->mod_shm.addr + shm_mod->name, xpath_sub[del_idx2].cid);
         evpipe_num = xpath_sub[del_idx2].evpipe_num;
 
         /* remove the subscription */
@@ -1156,7 +1148,7 @@ sr_shmext_oper_get_sub_stop(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t del
         }
     }
 
-    if (has_locks == SR_LOCK_READ) {
+    if (has_locks == SR_LOCK_NONE) {
         /* EXT READ UNLOCK */
         sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 1, __func__);
 
@@ -1165,11 +1157,6 @@ sr_shmext_oper_get_sub_stop(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t del
                 __func__, NULL, NULL))) {
             sr_errinfo_merge(&err_info, tmp_err);
         }
-
-        /* EXT READ LOCK */
-        if ((tmp_err = sr_shmext_conn_remap_lock(conn, SR_LOCK_READ, 0, __func__))) {
-            sr_errinfo_merge(&err_info, tmp_err);
-        }
     }
 
     if (del_evpipe) {
@@ -1926,7 +1913,7 @@ sr_shmext_recover_sub_oper_get(sr_conn_ctx_t *conn, sr_mod_t *shm_mod)
         shm_sub = &((sr_mod_oper_get_sub_t *)(conn->ext_shm.addr + shm_mod->oper_get_subs))[i];
 
         for (count = shm_sub->xpath_sub_count; count; --count) {
-            if ((tmp_err = sr_shmext_oper_get_sub_stop(conn, shm_mod, i, count - 1, 1, SR_LOCK_WRITE, 1))) {
+            if ((tmp_err = sr_shmext_oper_get_sub_stop(conn, shm_mod, i, count - 1, SR_LOCK_WRITE, 0))) {
                 sr_errinfo_merge(&err_info, tmp_err);
             }
         }
diff --git a/src/shm_ext.h b/src/shm_ext.h
index 3a1c368..df3503c 100644
--- a/src/shm_ext.h
+++ b/src/shm_ext.h
@@ -150,13 +150,12 @@ sr_error_info_t *sr_shmext_oper_get_sub_del(sr_conn_ctx_t *conn, sr_mod_t *shm_m
  * @param[in] shm_mod SHM module with subscriptions.
  * @param[in] del_idx1 Index of the list of subscriptions with the same XPath.
  * @param[in] del_idx2 Index of the XPath subscription to free.
- * @param[in] del_evpipe Whether to also remove the evpipe.
- * @param[in] has_locks Mode of held CHANGE SUB and EXT locks.
- * @param[in] recovery Whether to print subscription recovery warning.
+ * @param[in] has_locks Mode of held EXT lock.
+ * @param[in] sub_id id to delete if locks are not held.
  * @return err_info, NULL on success.
  */
 sr_error_info_t *sr_shmext_oper_get_sub_stop(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, uint32_t del_idx1, uint32_t del_idx2,
-        int del_evpipe, sr_lock_mode_t has_locks, int recovery);
+        sr_lock_mode_t has_locks, uint32_t sub_id);
 
 /**
  * @brief Add main SHM module operational poll subscription.
diff --git a/src/shm_sub.c b/src/shm_sub.c
index c8acfbb..01ec945 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -367,6 +367,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
         ret = sr_cond_clockwait(&sub_shm->lock.cond, &sub_shm->lock.mutex, COMPAT_CLOCK_ID, &timeout_abs);
     }
 
+    last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
     if (!sub_shm->lock.readers[0]) {
         /* FAKE WRITE LOCK */
         assert(!sub_shm->lock.writer);
@@ -374,15 +375,17 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
 
         if (ret == ETIMEDOUT) {
             /* try to recover the event again in case the originator crashed later */
-            sr_shmsub_recover(sub_shm);
-            if (!ATOMIC_LOAD_RELAXED(sub_shm->event)) {
-                /* recovered */
+            if (last_event) {
+                sr_shmsub_recover(sub_shm);
+                last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
+            }
+            if (!last_event) {
+                /* recovered or the originator cleared the event just as we see a timeout */
                 ret = 0;
             }
         }
     }
 
-    last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
     last_request_id = ATOMIC_LOAD_RELAXED(sub_shm->request_id);
 
     if (ret) {
@@ -500,6 +503,7 @@ _sr_shmsub_notify_wait_wr(sr_sub_shm_t *sub_shm, sr_sub_event_t event, uint32_t
             /* event failed */
             if (clear_ev_on_err) {
                 ATOMIC_STORE_RELAXED(sub_shm->event, SR_SUB_EV_NONE);
+                SR_LOG_INF("Clearing event ID %" PRIu32 " after timeout", request_id);
             } else if ((expected_ev == SR_SUB_EV_SUCCESS) || (expected_ev == SR_SUB_EV_ERROR)) {
                 ATOMIC_STORE_RELAXED(sub_shm->event, SR_SUB_EV_ERROR);
             }
@@ -2078,7 +2082,7 @@ cleanup:
 
 sr_error_info_t *
 sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, const char *request_xpath,
-        const struct lyd_node *parent, const char *orig_name, const void *orig_data, sr_mod_oper_get_sub_t *oper_get_subs,
+        const struct lyd_node *parent, const char *orig_name, const void *orig_data, sr_mod_oper_get_sub_cached_t *cached_subs,
         uint32_t idx1, uint32_t timeout_ms, sr_conn_ctx_t *conn, struct lyd_node **data, sr_error_info_t **cb_err_info)
 {
     sr_error_info_t *err_info = NULL;
@@ -2088,6 +2092,7 @@ sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, cons
     char *parent_lyb = NULL;
     struct lyd_node *oper_data;
     sr_cid_t cid;
+    uint32_t cached_sub_count = cached_subs[idx1].xpath_sub_count;
 
     if (!request_xpath) {
         request_xpath = "";
@@ -2095,13 +2100,13 @@ sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, cons
     cid = conn->cid;
 
     i = 0;
-    while (i < oper_get_subs[idx1].xpath_sub_count) {
-        xpath_sub = &((sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + oper_get_subs[idx1].xpath_subs))[i];
+    while (i < cached_sub_count) {
+        xpath_sub = &(cached_subs[idx1].xpath_subs[i]);
 
         /* check subscription aliveness */
         if (!sr_conn_is_alive(xpath_sub->cid)) {
             /* recover the subscription */
-            if ((err_info = sr_shmext_oper_get_sub_stop(conn, mod->shm_mod, idx1, i, 1, SR_LOCK_READ, 1))) {
+            if ((err_info = sr_shmext_oper_get_sub_stop(conn, mod->shm_mod, idx1, i, SR_LOCK_NONE, xpath_sub->sub_id))) {
                 sr_errinfo_free(&err_info);
             }
 
@@ -2109,6 +2114,8 @@ sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, cons
             if ((err_info = sr_shmsub_oper_poll_get_sub_change_notify_evpipe(conn, mod->ly_mod->name, xpath))) {
                 sr_errinfo_free(&err_info);
             }
+            /* these are cached subs, we must advance as stop only fixes the shm */
+            i++;
             continue;
         }
 
diff --git a/src/shm_sub.h b/src/shm_sub.h
index 76cbb11..823c8ae 100644
--- a/src/shm_sub.h
+++ b/src/shm_sub.h
@@ -176,7 +176,7 @@ sr_error_info_t *sr_shmsub_change_notify_change_abort(struct sr_mod_info_s *mod_
  * @return err_info, NULL on success.
  */
 sr_error_info_t *sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, const char *request_xpath,
-        const struct lyd_node *parent, const char *orig_name, const void *orig_data, sr_mod_oper_get_sub_t *oper_get_subs,
+        const struct lyd_node *parent, const char *orig_name, const void *orig_data, sr_mod_oper_get_sub_cached_t *oper_get_subs,
         uint32_t idx1, uint32_t timeout_ms, sr_conn_ctx_t *conn, struct lyd_node **data, sr_error_info_t **cb_err_info);
 
 /**
diff --git a/src/shm_types.h b/src/shm_types.h
index 04a55a1..798a4c0 100644
--- a/src/shm_types.h
+++ b/src/shm_types.h
@@ -236,6 +236,15 @@ typedef struct {
     uint32_t xpath_sub_count;   /**< Number of subscriptions for given XPath */
 } sr_mod_oper_get_sub_t;
 
+/**
+ * @brief cached oper get subscriptions so we don't need to hold remap lock
+ * */
+typedef struct {
+    sr_mod_oper_get_xpath_sub_t *xpath_subs; /**< XPath sub array */
+    uint32_t xpath_sub_count;   /**< Number of subscriptions for given XPath */
+    char *xpath;                /**< XPath of the subscription */
+} sr_mod_oper_get_sub_cached_t;
+
 /**
  * @brief Ext SHM module operational poll subscription.
  */
